### PR TITLE
TASK: Property typing for policy classes

### PR DESCRIPTION
**Pull request by daniellienert**

### CODE REVIEW

**Review by sorenmalling — APPROVED** [review]
Looks good from reading on mobile

File: Neos.Flow/Classes/Security/Context.php [hunk 1]
```diff
@@ -396,6 +396,7 @@ public function getAuthenticationTokensOfType($className)
      * @return Role[]
      * @throws Exception
      * @throws Exception\NoSuchRoleException
+     * @throws \Neos\Flow\Configuration\Exception\InvalidConfigurationTypeException
```
Comment (bwaidelich) [inline Neos.Flow/Classes/Security/Context.php:399]: nitpick: import namespace?

File: Neos.Flow/Classes/Security/Policy/PolicyService.php [hunk 2]
```diff
@@ -279,11 +292,13 @@ public function getPrivilegeTargets()
      *
      * @param string $privilegeTargetIdentifier Identifier of a privilege target
      * @return PrivilegeTarget
```
Comment (bwaidelich) [inline Neos.Flow/Classes/Security/Policy/PolicyService.php:294]:
```suggestion
     * @return ?PrivilegeTarget
```

File: Neos.Flow/Classes/Security/Policy/Role.php [hunk 3]
```diff
@@ -19,7 +21,7 @@
  */
 class Role
 {
-    const ROLE_IDENTIFIER_PATTERN = '/^(\w+(?:\.\w+)*)\:(\w+)$/';   // Vendor(.Package)?:RoleName
+    protected const ROLE_IDENTIFIER_PATTERN = '/^(\w+(?:\.\w+)*)\:(\w+)$/';   // Vendor(.Package)?:RoleName
```
Comment (bwaidelich) [inline Neos.Flow/Classes/Security/Policy/Role.php:24]: I'd say. Let's make the class `final` and this constant `private`

**Review by bwaidelich — COMMENTED** [review]
Nice one, thanks for taking care!
Just a few (mostly nitpicky) comments

File: Neos.Flow/Classes/Security/Policy/PolicyService.php (hunk 2 above)
Comment (albe) [reply Neos.Flow/Classes/Security/Policy/PolicyService.php:294]: Do we actually use the `?` optional type prefix in docblocks? IIRC our current reflection does not parse that correctly - the alternative is `|null`

File: Neos.Flow/Classes/Security/Policy/Role.php (hunk 3 above)
Comment (albe) [reply Neos.Flow/Classes/Security/Policy/Role.php:24]: Devil's advocat: Making that class final would have to be considered breaking and be documented clearly. Let's maybe do this in #1939 instead?

File: Neos.Flow/Classes/Security/Policy/PolicyService.php (hunk 2 above)
Comment (bwaidelich) [reply Neos.Flow/Classes/Security/Policy/PolicyService.php]: yes, you're right. Updated the suggestion!

File: Neos.Flow/Classes/Security/Policy/Role.php (hunk 3 above)
Comment (bwaidelich) [reply Neos.Flow/Classes/Security/Policy/Role.php]: Yep, good point.

**Comment by daniellienert** [comment]
@bwaidelich, sorry had to remove the `final` from the policy class again, as we mock it in tests. Already this task consumed a lot of time and I would like to merge it as is.

**Pull request merged**